File: LeetCode/173BinarySearchTreeIterator/BinarySearchTreeIterator.py
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        # Iterative DFS -> Time:O(n), Space:O(h)
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        res = self.stack.pop()
        curr = res.right
        while curr:
            self.stack.append(curr)
            curr = curr.left
        return res.val

    def hasNext(self) -> bool:
        return self.stack != []

File: LeetCode/173BinarySearchTreeIterator/BinarySearchTreeIterator.cpp
class BSTIterator {
public:
    stack<TreeNode*> s;
    BSTIterator(TreeNode* root) {
        // Iterative DFS -> Time:O(n), Space:O(h)
        while (root) {
            s.push(root);
            root = root->left;
        }
    }
    
    int next() {
        TreeNode *res, *curr;
        res = s.top();
        s.pop();
        curr = res->right;
        while (curr) {
            s.push(curr);
            curr = curr->left;
        }
        return res->val;
    }
    
    bool hasNext() {
        return !s.empty();
    }
};
